File: data-stuctures/Queue.py
class Node:
    def __init__(self, val: int, next=None):
        self.val = val
        self.next = next

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def enqueue(self, val: int):
        node = Node(val=val)
        self.length += 1
        if not self.tail:
            self.tail = node
            self.head = self.tail
        else:
            self.tail.next = node
            self.tail = node
    
    def dequeue(self):
        if not self.head:
            return None
        else:
            self.length -= 1
            tempHead = self.head
            self.head = self.head.next
            tempHead.next = None
            if self.length == 0:
                self.tail = None
            return tempHead.val
    
    def peek(self):
        if self.head:
            return self.head.val
        return None


que = Queue()

print(que.dequeue())
print(que.length)
que.enqueue(5)
print(que.dequeue())
print(que.length)
que.enqueue(5)
que.enqueue(7)
que.enqueue(2)
print(que.length)
print(que.peek())
print(que.dequeue())
print(que.dequeue())
print(que.length)
print(que.peek())
print(que.dequeue())
print(que.peek())

File: data-stuctures/dart/Queue.dart
class Node<T> {
  T value;
  Node<T>? next;

  Node(this.value);
}

class Queue<T> {
  int _length = 0;

  Node<T>? _head;

  Node<T>? _tail;

  Queue() {
    this._head = null;
    this._tail = null;
  }

  int get length => _length;

  void enqueue(T item) {
    _length++;
    Node<T> node = Node(item);
    if (_tail != null) {
      _tail?.next = node;
      _tail = node;
    } else {
      _tail = node;
      _head = _tail;
    }
  }

  T? deque() {
    if (_head == null) {
      return null;
    } else {
      this._length--;
      Node<T> head = _head!;
      _head = _head?.next;
      head.next = null;
      if (_length == 0) {
        _tail = null;
      }
      return head.value;
    }
  }

  T? peek() {
    return _head?.value;
  }
}

void main(List<String> args) {
  Queue<int> queue = Queue();

  queue.enqueue(5);

  print(queue.length);

  queue.enqueue(7);

  queue.enqueue(12);

  print(queue.length);

  print(queue.deque());

  print(queue.peek());

  print(queue.deque());

  print(queue.peek());

  print(queue.deque());

  print(queue.deque());

  print(queue.peek());
}
